typegen/src/utils.rs

```diff
@@ -81,25 +82,50 @@ pub fn ensure_unique_type_paths(types: &mut PortableRegistry) {
 /// all type ids mentioned in the TypeDef are either:
 /// - equal
 /// - or different, but map essentially to the same generic type parameter
-fn types_equal_extended_to_params(a: &Type<PortableForm>, b: &Type<PortableForm>) -> bool {
-    let collect_params = |type_params: &[TypeParameter<PortableForm>]| {
-        type_params
-            .iter()
-            .filter_map(|p| p.ty.map(|ty| (ty.id, p.name.clone())))
-            .collect::<HashMap<u32, String>>()
-    };
+pub(crate) fn types_equal_extended_to_params(
+    a: &Type<PortableForm>,
+    b: &Type<PortableForm>,
+) -> bool {
+    // We map each type ID to all type params if could refer to. Type IDs can refer to multiple parameters:
+    // E.g. Foo<A,B> can be parameterized as Foo<u8,u8>, so if 42 is the type id of u8, a filed with id=42 could refer to eiher A or B.
+    fn collect_params(
+        type_params: &[TypeParameter<PortableForm>],
+    ) -> HashMap<u32, SmallVec<[&str; 2]>> {
+        let mut hm: HashMap<u32, SmallVec<[&str; 2]>> = HashMap::new();
+        for p in type_params {
+            if let Some(ty) = &p.ty {
+                match hm.entry(ty.id) {
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().push(p.name.as_str());
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(smallvec![p.name.as_str()]);
+                    }
+                }
+            }
+        }
+        hm
+    }
 
     let type_params_a = collect_params(&a.type_params);
-    let type_params_b = collect_params(&a.type_params);
+    let type_params_b = collect_params(&b.type_params);
 
-    if type_params_a.len() != type_params_b.len() {
+    if a.type_params.len() != b.type_params.len() {
         return false;
     }
-
-    // returns true if the ids are the same OR if they point to the same generic parameter.
+    // Returns true if the ids are the same OR if they point to the same generic parameter.
     let ids_equal = |a: u32, b: u32| -> bool {
-        a == b
-            || matches!((type_params_a.get(&a), type_params_b.get(&b)), (Some(a_name), Some(b_name)) if a_name == b_name)
+        if a == b {
+            return true;
+        }
+        let Some(a_param_names) = type_params_a.get(&a) else {
+            return false;
+        };
+        let Some(b_param_names) = type_params_b.get(&b) else {
+            return false;
+        };
+        // Check if there is any intersection, meaning that both IDs map to the same generic type param:
+        a_param_names.iter().any(|a_p| b_param_names.contains(a_p))
```

Comment:
So this will treat `struct B<usize, String, bool>  == struct A<(), (), usize>` because both contains `usize`? Or am I missing something?

Reply:
seems like like `a.name == b.name` is checked before this is called on composite/types should fine I think :)

Reply:
No, it is a bit more subtle: This `ids_equal` function is called for every field of the type: Let's say we have a generic type like this:
```
struct A<T,U,V> { 
    f1: T,
    f2: U,
    f3: V
}
```
Let's say the concrete types `A<usize, String, bool>` and `A<(), (), usize>` have been registered in the type registry and the ids of the primitive types are: `usize: id=0`, `String: id=1`, `bool: id=2`, `(): id=3`. Then `A<usize, String, bool>` is represented as:
```
NamedComposite<T: id=0, U: id=1, V: id=2> {
    f1: id=0,
    f2: id=1,
    f3: id=2
}
```
and `A<(), (), usize>` is represented as 
```
NamedComposite<T: id=3, U: id=3, V: id=0> {
    f1: id=3,
    f2: id=3,
    f3: id=0
}
```
Now we compare  `A<usize, String, bool>` and `A<(), (), usize>`  to find out if they are basically the same rust type (they should be, both are parameterizations of a generic `A<T,U,V>`). 
Now for field f1, we see that the type ids do not equal (`0 != 3`). But: `a_param_names` will say, oh type id=0 actually maps to `[T]` and `b_param_names` will say, oh type_id 3 maps to `[T, or U]`.
Now the `a_param_names.iter().any(|a_p| b_param_names.contains(a_p))` returns true, saying "okay this field is just the generic parameter `T` for both registered types, so they are considered the same".
Then we continue with the next field `f2` and so on.